staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler_test.go

```diff
@@ -418,3 +439,246 @@ func TestRouting(t *testing.T) {
+		} else if item.GetObjectKind().GroupVersionKind() != expectGVK {
+			t.Errorf("expected create result to be %#v, got %#v", expectGVK, item.GetObjectKind().GroupVersionKind())
+		} else {
+			u = item.(*unstructured.Unstructured)
+		}
+
+		// Update
+		u.SetAnnotations(map[string]string{"updated": "true"})
+		if item, _, err := crdInfo.storages[version.Name].CustomResource.Update(context.TODO(), u.GetName(), rest.DefaultUpdatedObjectInfo(u), validateFunc, updateValidateFunc, false, &metav1.UpdateOptions{}); err != nil {
+			t.Fatal(err)
+		} else if item.GetObjectKind().GroupVersionKind() != expectGVK {
+			t.Errorf("expected update result to be %#v, got %#v", expectGVK, item.GetObjectKind().GroupVersionKind())
+		}
+
+		// Get
+		if item, err := crdInfo.storages[version.Name].CustomResource.Get(context.TODO(), u.GetName(), &metav1.GetOptions{}); err != nil {
+			t.Fatal(err)
+		} else if item.GetObjectKind().GroupVersionKind() != expectGVK {
+			t.Errorf("expected get result to be %#v, got %#v", expectGVK, item.GetObjectKind().GroupVersionKind())
+		}
+
+		if enableWatchCache {
+			// Allow time to propagate the create into the cache
+			time.Sleep(time.Second)
+			// Get cached
+			if item, err := crdInfo.storages[version.Name].CustomResource.Get(context.TODO(), u.GetName(), &metav1.GetOptions{ResourceVersion: "0"}); err != nil {
+				t.Fatal(err)
+			} else if item.GetObjectKind().GroupVersionKind() != expectGVK {
+				t.Errorf("expected cached get result to be %#v, got %#v", expectGVK, item.GetObjectKind().GroupVersionKind())
+			}
+		}
+	}
+
+	// List every version, expect all returned items to match request GVK
+	for _, version := range crd.Spec.Versions {
+		expectGVK := schema.GroupVersionKind{Group: "stable.example.com", Version: version.Name, Kind: "MultiVersion"}
+
+		if list, err := crdInfo.storages[version.Name].CustomResource.List(context.TODO(), &metainternalversion.ListOptions{}); err != nil {
+			t.Fatal(err)
+		} else {
+			for _, item := range list.(*unstructured.UnstructuredList).Items {
+				if item.GroupVersionKind() != expectGVK {
+					t.Errorf("expected list item to be %#v, got %#v", expectGVK, item.GroupVersionKind())
+				}
+			}
+		}
+
+		if enableWatchCache {
+			// List from watch cache
+			if list, err := crdInfo.storages[version.Name].CustomResource.List(context.TODO(), &metainternalversion.ListOptions{ResourceVersion: "0"}); err != nil {
+				t.Fatal(err)
+			} else {
+				for _, item := range list.(*unstructured.UnstructuredList).Items {
+					if item.GroupVersionKind() != expectGVK {
+						t.Errorf("expected cached list item to be %#v, got %#v", expectGVK, item.GroupVersionKind())
+					}
+				}
+			}
+		}
+
+		watch, err := crdInfo.storages[version.Name].CustomResource.Watch(context.TODO(), &metainternalversion.ListOptions{ResourceVersion: startResourceVersion})
+		if err != nil {
+			t.Fatal(err)
+		}
+		// 5 events: delete marker, create v1alpha1, create v1beta1, update v1alpha1, update v1beta1
```

Comment:
I assume etcd compaction is a non-issue?

Reply:
yeah, compaction runs every 5 minutes. this test takes ~2-5 seconds